src/odm_sharing/private/queries.py

```diff
@@ -1,14 +1,326 @@
+        '''associate (partial) sql query with node's rule'''
+        assert node.kind in [NodeKind.TABLE, NodeKind.FILTER, NodeKind.GROUP]
+        rule_queries[node.rule_id] = PartialQuery(sql=sql, args=args)
+
+    n = node
+    if n.kind == NodeKind.TABLE:
+        # table has a select followed by an optional filter/group
+        table = n.str_val
+        select = n.sons[0]
+        filter_root = n.sons[1] if len(n.sons) > 1 else None
+        select_sql = f'SELECT {recurse(select)} FROM {ident(table)}'
+        filter_sql = f'WHERE {recurse(filter_root)}' if filter_root else ''
+        record(n, select_sql, [])
+        return (select_sql + ' ' + filter_sql)
+    elif n.kind == NodeKind.SELECT:
+        # select has an all-value or fields as children
+        if n.str_val == ALL_LIT:
+            return '*'
+        else:
+            columns = seq(n.sons).map(lambda x: ident(x.str_val))
+            return ','.join(columns)
+    elif n.kind == NodeKind.GROUP:
+        # group combines any number of children with its operator
+        op_str = n.str_val.upper()
+        assert op_str in ['AND', 'OR']
+        arg_start = len(args)
+        sql = seq(n.sons)\
+            .map(recurse)\
+            .reduce(lambda x, y: f'({x} {op_str} {y})')
+        record(n, sql, args[arg_start:])
+        return sql
+    elif n.kind == NodeKind.FILTER:
+        # filter has op as value, children define field, kind and literals
+        op = parse_op(n.str_val)
+        key_ident = ident(recurse(n.sons[0]))
+
+        def gen_range_sql(range_kind: RangeKind, values: List[str]) -> str:
+            '''generates sql for a range of values'''
+            if range_kind == RangeKind.INTERVAL:
+                a = values[0]
+                b = values[1]
+                return f'({key_ident} BETWEEN {a} AND {b})'
+            elif range_kind == RangeKind.SET:
+                return f"({key_ident} IN ({','.join(values)}))"
+            else:
+                assert False, "unreachable"
+                return ''
+
+        if op == Op.RANGE:
+            range_kind = RangeKind(n.sons[1].str_val)
+            assert range_kind != RangeKind.SINGLE
+            literals: List[str] = seq(n.sons[2:]).map(recurse).list()
+            sql = gen_range_sql(range_kind, literals)
+            record(n, sql, args[-len(literals):])
+            return sql
+        else:
+            val = recurse(n.sons[1])
+            sql = f'({key_ident} {op} {val})'
+            record(n, sql, args[-1:])
+            return sql
+    elif n.kind == NodeKind.FIELD:
+        # field name is quoted
+        return dqt(n.str_val)
+    elif n.kind == NodeKind.LITERAL:
+        # literal value is converted and added to list of args, while only a
+        # parameter placeholder is added to the sql
+        val = convert(n.str_val)
+        args.append(val)
+        return '?'
+    else:
+        assert False, 'unreachable'
+        return ''
+
+
+def gen_sql(
+    root_node: Node,
```

Comment:
> I'm assuming that this is not the root of the tree but a table node? If that's the case can you change the name to `root_table_node` or `table_node`?

it's the root of the sub-tree that is passed in, but you're right that it should be called `table_node` instead. Fixed.

> Can we rename this to `gen_data_sql`? So that it's not confused with generating the count sql?

fixed